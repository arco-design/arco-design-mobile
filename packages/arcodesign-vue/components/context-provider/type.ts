export interface ContextInstallOptions {
    /**
     * 组件名称前缀
     * @en Component prefix
     * @default "Arco"
     */
    componentPrefix?: string;
}

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
}
