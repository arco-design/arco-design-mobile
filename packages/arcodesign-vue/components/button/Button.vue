<template>
    <button
        :class="[
            prefixCls,
            `${prefixCls}-size-${size}`,
            `${prefixCls}-${type}`,
            `${prefixCls}-shape-${shape}`,
            {
                'is-inline': inline,
                'is-loading': loading,
                'is-disabled': disabled || (disableWhenLoading && loading),
                'is-active': isActive,
            },
            ...customClass,
        ]"
        :style="buttonStyle"
        @touchstart="handleTouchStart"
        @touchend="handleTouchEnd"
        @touchcancel="handleTouchEnd"
        @click="handleClick"
    >
        <span v-if="icon" :class="`${prefixCls}-icon`">
            <slot name="icon">{{ icon }}</slot>
        </span>
        <span v-if="!loading || showTextWhenLoading" :class="`${prefixCls}-text`">
            <slot>{{ children }}</slot>
        </span>
        <!-- Display the loading icon if loading is true -->
        <loading v-if="loading" :class="`${prefixCls}-loading-icon`" />
    </button>
</template>

<script lang="ts">
import { defineComponent, computed, ref, onMounted, watch } from 'vue';
import { ButtonProps, buttonProps } from './type';
import useButtonClass from './hooks';
import Loading from '@/loading';

export default defineComponent({
    name: 'ArcButton',
    components: { Loading },
    props: buttonProps,
    emits: ['click'],
    setup(props: ButtonProps, { emit, slots }) {
        const prefixCls = 'arc-button';
        const isActive = ref(false);

        const handleTouchStart = () => {
            if (props.needActive && !props.disabled) {
                isActive.value = true;
            }
        };

        const handleTouchEnd = () => {
            isActive.value = false;
        };

        const handleClick = (event: Event) => {
            if (props.disabled) {
                props.onClickDisabled?.(event);
                return;
            }
            if (!props.loading || !props.disableWhenLoading) {
                props.onClick?.(event);
            }
        };

        const { buttonClass, buttonStyle } = useButtonClass(props);

        watch(
            () => props.loading,
            newLoading => {
                if (!newLoading) {
                    isActive.value = false;
                }
            },
        );

        onMounted(() => {
            // Handle additional lifecycle logic if needed
        });

        return {
            prefixCls,
            buttonClass,
            buttonStyle,
            isActive,
            handleTouchStart,
            handleTouchEnd,
            handleClick,
        };
    },
});
</script>
