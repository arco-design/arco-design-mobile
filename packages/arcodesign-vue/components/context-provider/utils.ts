import { InjectionKey, inject } from 'vue';
import { ContextInstallOptions, GlobalContextParams } from './type';

export const defaultInstallOptions: Required<ContextInstallOptions> = {
    componentPrefix: 'Arco',
};

export const defaultContext: Required<GlobalContextParams> = {
    prefixCls: 'arco',
};

export const configProviderInjectionKey: InjectionKey<GlobalContextParams> =
    Symbol('ArcoContextProvider');

export function getComponentPrefix(options?: ContextInstallOptions) {
    return options?.componentPrefix || defaultInstallOptions.componentPrefix;
}

export function getPrefixCls(componentName?: string) {
    const contextProvider = inject(configProviderInjectionKey, defaultContext);
    const prefix = contextProvider.prefixCls;
    if (componentName) {
        return `${prefix}-${componentName}`;
    }
    return prefix;
}
