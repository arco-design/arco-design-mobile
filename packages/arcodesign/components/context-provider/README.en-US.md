### Others

# ContextProvider 

Global data control component, used to replace global data.

======

> Props

|Property|Description|Type|DefaultValue|
|----------|-------------|------|------|
|prefixCls|Component classname prefix|string|"arco"|
|system|Manually control the current system, and the incoming value will be used directly after being passed in\. It is applicable when the initial value of the system needs to be specified in the ssr scenario\.|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|Whether to use dark mode|boolean|false|
|isDarkMode|Whether it is in dark mode|boolean|false|
|theme|Theme variable\. The css variable will be replaced online after input\. The less variable needs to be set|Record\<string, string\>|-|
|locale|Internationalized language configuration|ILocale|-|
|useRtl|Whether to use rtl|boolean|false|

> ILocale

|Property|Description|Type|
|----------|-------------|------|
|locale|Language Type|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
