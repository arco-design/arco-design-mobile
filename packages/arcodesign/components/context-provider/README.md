### 其他

# 全局配置 ContextProvider

全局数据控制组件，用于替换全局数据。

======

> 属性/Props

|参数|描述|类型|默认值|
|----------|-------------|------|------|
|prefixCls|组件类名前缀|string|"arco"|
|system|手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|是否使用暗黑模式|boolean|false|
|isDarkMode|是否处于暗黑模式，指定后以指定的值为准|boolean|false|
|theme|主题变量，传入后将在线替换css变量，需设置less变量 @use\-css\-vars: 1|Record\<string, string\>|-|
|locale|国际化语言包配置|ILocale|-|
|useRtl|是否使用Rtl模式|boolean|false|

> ILocale

|参数|描述|类型|
|----------|-------------|------|
|locale|语言类型|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
