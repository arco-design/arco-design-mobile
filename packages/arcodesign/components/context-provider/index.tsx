import React, { createContext, useState, useEffect, useCallback } from 'react';
import { addCssRules, removeCssStyleDom, defaultLocale, ILocale } from '@arco-design/mobile-utils';

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
    /**
     * 手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用
     * @en Manually control the current system, and the incoming value will be used directly after being passed in. It is applicable when the initial value of the system needs to be specified in the ssr scenario.
     * @default ""
     */
    system?: 'pc' | 'android' | 'ios' | '';
    /**
     * 是否使用暗黑模式
     * @en Whether to use dark mode
     * @default false
     */
    useDarkMode?: boolean;
    /**
     * 是否处于暗黑模式，指定后以指定的值为准
     * @en Whether it is in dark mode
     * @default false
     */
    isDarkMode?: boolean;
    /**
     * 主题变量，传入后将在线替换css变量，需设置less变量 @use-css-vars: 1
     * @en Theme variable. The css variable will be replaced online after input. The less variable needs to be set @use-css-vars: 1
     */
    theme?: Record<string, string>;
    /**
     * 国际化语言包配置
     * @en Internationalized language configuration
     */
    locale?: ILocale;
    /**
     * 是否使用Rtl模式
     * @en Whether to use rtl
     * @default false
     */
    useRtl?: boolean;
}

export const defaultContext: GlobalContextParams = {
    prefixCls: 'arco',
    system: '',
    useDarkMode: false,
    isDarkMode: false,
    locale: defaultLocale,
    useRtl: false,
};

export const GlobalContext = createContext<GlobalContextParams>(defaultContext);

export interface ContextProviderProps extends GlobalContextParams {
    children: React.ReactNode;
}

export type WithGlobalContext<T> = T & { context?: GlobalContextParams };

/**
 * 全局数据控制组件，用于替换全局数据。
 * @en Global data control component, used to replace global data.
 * @type 其他
 * @type_en Others
 * @name 全局配置
 * @name_en ContextProvider
 */
export default function ContextProvider(props: ContextProviderProps) {
    const {
        children,
        useDarkMode,
        isDarkMode,
        theme,
        locale = defaultLocale,
        ...restProps
    } = props;
    const [isDarkModeState, setIsDarkModeState] = useState<boolean>();
    const changeDarkMode = useCallback(
        (res: MediaQueryListEvent) => setIsDarkModeState(res.matches),
        [],
    );

    useEffect(() => {
        const matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
        if (useDarkMode) {
            const dark = matchMedia.matches;
            setIsDarkModeState(dark);
            if (typeof matchMedia.addEventListener === 'function') {
                matchMedia.addEventListener('change', changeDarkMode);
            } else if (typeof matchMedia.addListener === 'function') {
                matchMedia.addListener(changeDarkMode);
            }
        }
        return () => {
            if (useDarkMode) {
                if (typeof matchMedia.removeEventListener === 'function') {
                    matchMedia.removeEventListener('change', changeDarkMode);
                } else if (typeof matchMedia.removeListener === 'function') {
                    matchMedia.removeListener(changeDarkMode);
                }
            }
        };
    }, [useDarkMode]);

    useEffect(() => {
        if (!theme) {
            return;
        }
        addCssRules('arcoCustomTheme', theme);
        return () => {
            removeCssStyleDom('arcoCustomTheme');
        };
    }, [theme]);

    return (
        <GlobalContext.Provider
            value={{
                ...defaultContext,
                ...restProps,
                isDarkMode: isDarkMode === void 0 ? isDarkModeState : isDarkMode,
                useDarkMode,
                locale: locale || defaultLocale,
            }}
        >
            {children}
        </GlobalContext.Provider>
    );
}

export const ContextLayout = GlobalContext.Consumer;

export function CompWithGlobalContext<P extends JSX.IntrinsicAttributes>(
    Component: React.FunctionComponent<P>,
) {
    return function (props: WithGlobalContext<P>) {
        const { context: propsContext, ...others } = props;
        return (
            <ContextLayout>
                {context => (
                    <ContextProvider {...{ ...context, ...propsContext }}>
                        <Component {...(others as P)} />
                    </ContextProvider>
                )}
            </ContextLayout>
        );
    };
}
