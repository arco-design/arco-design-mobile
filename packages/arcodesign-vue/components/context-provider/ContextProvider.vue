<script lang="ts" setup>
import { provide } from 'vue';
import { GlobalContextParams } from './type';
import { configProviderInjectionKey, defaultContext } from './utils';

const props = withDefaults(defineProps<GlobalContextParams>(), defaultContext);
provide(configProviderInjectionKey, props);
</script>
<template>
    <slot />
</template>
