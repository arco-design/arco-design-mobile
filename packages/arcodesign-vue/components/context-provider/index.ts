import ContextProvider from './ContextProvider.vue';
import { installFactory } from '../_helpers/install';

export type * from './type';
export * from './utils';

const install = installFactory(ContextProvider);

export { ContextProvider };

export default {
    install,
};
