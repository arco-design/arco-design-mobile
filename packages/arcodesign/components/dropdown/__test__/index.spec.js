import React from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import Dropdown from '..';
import { mount } from 'enzyme';
import Button from '../../button';
import { defaultContext } from '../../context-provider';

const prefix = `${defaultContext.prefixCls}-dropdown`;

demoTest('dropdown');

mountTest(Dropdown, 'Dropdown');

const options= [
    {
        label: 'title1',
        value: 0,
        disabled: false,
    },
    {
        label: 'title2',
        value: 1,
    },
    {
        label: 'title3',
        value: 2,
        disabled: true,
    }
]

class Test extends React.Component {
    state = {
        value: false,
    };
    render() {
        const { value } = this.state;
        return (
            <div>
                <Button onClick={() => this.setState({
                    value: !value
                })}>Click Me</Button>
                <Dropdown options={options} onOptionClick={() => this.setState({value: !value})} touchToClose={false} showDropdown={value} {...this.props}/>
            </div>
        );
    }
}

describe('Dropdown style', () => {

    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        jest.useRealTimers();
    });

    it('Multi-column styles render correctly', () => {
        const wrapper = mount(
            <Dropdown
                useColumn={3}
                multiple={true}
                defaultSelectedValue={[0]}
                options={options}
                showDropdown={true}
                onOptionClick={() => { console.info('click'); }}
                onOptionChange={(value, item) => {
                    console.info(value,item);
                    // setShowDropdown(false);
                }}/>
        );
        wrapper.setProps({
            showDropdown: false,
            direction: 'up',
            height: '300px'
        })
        expect(wrapper.find(`.${prefix}-options-wrap.use-column`).length).toBe(1);
    });
    it('Custom nodes render correctly', () => {
        const wrapper = mount(
            <Dropdown
                showDropdown={true}
                getScrollContainer={() => document.getElementById('test')}
            >
                <div id="test">
                    content
                </div>
            </Dropdown>
        );
        expect(wrapper.find('#test').text()).toEqual('content');
    });
    it('Use options render correctly', () => {
        const wrapper = mount(
            <Dropdown
                showDropdown={true}
                direction="up">
                <Dropdown.Options
                    useColumn={3}
                    multiple={true}
                    selectedValue={[]}
                    options={options}
                ></Dropdown.Options>
                <Dropdown.Options
                    useColumn={3}
                    multiple={false}
                    selectedValue={[]}
                    options={[
                    {
                        label: 'title1',
                        value: 0,
                        disabled: false,
                    },
                    {
                        label: 'title2',
                        value: 1,
                    }]}
                ></Dropdown.Options>
            </Dropdown>
        );
        expect(wrapper.find(`.${prefix}-options-item-col`).length).toBe(5);
    });
})

describe('Dropdown open', () => {
    it('dropdown open correctly', () => {
        const wrapper = mount(<Test />);
        wrapper.find(Button).simulate('click');
        expect(wrapper.find(`.${prefix}-options-item`).length).toBe(3);
        wrapper.find(`.${prefix}-options-item`).at(1).simulate('click');
    });

    it('dropdown open in the right direction', () => {
        const wrapper = mount(<Test direction='up'/>);
        wrapper.find(Button).simulate('click');
        expect(wrapper.find('.drop-up').length).toBe(1);
    });
})
