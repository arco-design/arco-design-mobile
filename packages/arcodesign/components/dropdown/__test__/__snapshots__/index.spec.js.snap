// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown demo test dropdown demo: column.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-huge arco-button-size-huge-is-semi size-huge select-wrapper is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Click to expand
      </div>
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`dropdown demo test dropdown demo: custom-direction.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-huge arco-button-size-huge-is-semi size-huge select-wrapper is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Click to expand
      </div>
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`dropdown demo test dropdown demo: custom-node.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-huge arco-button-size-huge-is-semi size-huge select-wrapper is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Click to expand
      </div>
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`dropdown demo test dropdown demo: custom-options.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-huge arco-button-size-huge-is-semi size-huge select-wrapper is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Click to expand
      </div>
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`dropdown demo test dropdown demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <button
      class="arco-button arco-button-type-primary type-primary arco-button-size-huge arco-button-size-huge-is-semi size-huge select-wrapper is-semi pc"
      type="button"
    >
      <div
        class="arco-button-text arco-button-text-pc btn-text"
      >
        Click to expand
      </div>
    </button>
    <div />
  </div>
</DocumentFragment>
`;
